### PR TITLE
KEYCLOAK-12001: Audience support for SAML clients

### PR DESCRIPTION
**Pull request by rmartinc**
Adding audience mappers for SAML. I have replicated the same behavior that OIDC has (even the labels). Comments:

* I used the (not used yet in any mapper) SAMLLoginResponseMapper but I modified the interface to pass the ClientSessionContext instead of the AuthenticatedClientSessionModel. As it was not used by any mapper and it's not documented I decided to change the interface.
* I reused all the labels in the mappers from the OIDC (only needed  in one of them) and replicated the same behavior.
* I created an audience resolve mapper although I don't think it was totally needed, but I preferred to implement the same mapper that exist in OIDC.

@hmlnarik As always if you want some change or you see any mistake or improvement just let me now. As in the other SAML PR there is no hurry at all, I'm just completing this in order to give a better SAML assertion replay experience.

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
  */
 public interface SAMLLoginResponseMapper {
 
+
     ResponseType transformLoginResponse(ResponseType response, ProtocolMapperModel mappingModel, KeycloakSession session,
-                                        UserSessionModel userSession, AuthenticatedClientSessionModel clientSession);
+                                        UserSessionModel userSession, ClientSessionContext clientSessionCtx);
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java]: Why is this change here?

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thanks @rmartinc ! I realized I missed pushing "Submit review", sorry about that

File: services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLAudienceResolveProtocolMapper.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                    if (SamlProtocol.LOGIN_PROTOCOL.equals(app.getProtocol()) &&
+                            !app.getClientId().equals(clientSessionCtx.getClientSession().getClient().getClientId())) {
+                        audiences.add(app.getClientId());
+                    }
+                }
+            }
+            logger.debugf("Calculated audiences to add: %s", audiences);
+            // add the audiences
+            for (String audience : audiences) {
+                aud.addAudience(URI.create(audience));
```
[the remark this answers is not in the document]
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLAudienceResolveProtocolMapper.java]: I have added a try block in both mappers.

File: services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLAudienceProtocolMapper.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    public String getHelpText() {
+        return "Add specified audience to the audience conditions in the assertion.";
+    }
+
+    protected static AudienceRestrictionType locateAudienceRestriction(ResponseType response) {
+        try {
+            return response.getAssertions().get(0).getAssertion().getConditions().getConditions()
+                    .stream()
+                    .filter(c -> c instanceof AudienceRestrictionType)
```
[the remark this answers is not in the document]
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLAudienceProtocolMapper.java]: Done!

File: services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java (hunk 1 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java]: This change is because the audience resolve mapper needs to compute the groups of the user in each client (the resolve mapper duplicates the OIDC behavior, and the clientId is added in the audience if the user has at least one group in that client). So for that I re-used the [clientSessionCtx.getRoles()](https://github.com/keycloak/keycloak/pull/6520/files#diff-ec80ecf52bea4386f9a2e6e04c51d6b1R85) method. This class is also passed to the [SAMLRoleListMapper](https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLRoleListMapper.java#L32). If I hadn't passed it I would have needed to calculate groups again (repeating code and calculation), as I saw no mapper used this interface I decided to modify it (the previous AuthenticatedClientSessionModel can also be accessed through the ClientSessionContext).

**Comment by rmartinc** [comment]
Happy new year @hmlnarik! I added the first two modifications to the PR and explained why I changed the interface. Let me know if you prefer something different for the third point.

File: services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java (hunk 1 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/protocol/saml/mappers/SAMLLoginResponseMapper.java]: Thanks for explanation, it really makes sense to keep this on par with `OIDCAccessTokenMapper`.

**Review by hmlnarik — APPROVED** [review]

**Comment by iankko** [comment]
@rmartinc This one seems to have broken current _master_, when built with the distribution (_-Pdistribution_) profile. Filed [KEYCLOAK-12865](https://issues.redhat.com/browse/KEYCLOAK-12865) to track this. Will you have a chance to look at this yet && propose a fix, or should we fix it?

Thank you!, Jan

**Comment by iankko** [comment]
@rmartinc 

Created also [a PR for it](https://github.com/keycloak/keycloak/pull/6712) (seemed to be pretty straightforward). Could you review?

Thanks! Jan

> @rmartinc This one seems to have broken current _master_, when built with the distribution (_-Pdistribution_) profile. Filed [KEYCLOAK-12865](https://issues.redhat.com/browse/KEYCLOAK-12865) to track this. Will you have a chance to look at this yet && propose a fix, or should we fix it?
> 
> Thank you!, Jan

**Comment by rmartinc** [comment]
> @rmartinc
> 
> Created also [a PR for it](https://github.com/keycloak/keycloak/pull/6712) (seemed to be pretty straightforward). Could you review?
> 
> Thanks! Jan
> 
> > @rmartinc This one seems to have broken current _master_, when built with the distribution (_-Pdistribution_) profile. Filed [KEYCLOAK-12865](https://issues.redhat.com/browse/KEYCLOAK-12865) to track this. Will you have a chance to look at this yet && propose a fix, or should we fix it?
> > Thank you!, Jan

The PR LGTM. I have even run that test in my env. Wait for the tests to pass. I have approved your PR but I don't know if I can do that. :-P

**Comment by iankko** [comment]
@rmartinc 

> The PR LGTM. I have even run that test in my env. Wait for the tests to pass. I have approved your PR but I don't know if I can do that. :-P

Sure you can. Thanks for doing that & for re-running the necessary tests on your part! Just waiting for the Travis tests that one to finish OK too yet.
